src/terms.ml

```diff
@@ -10,6 +10,98 @@
 open Extra
 open Timed
 
+(** {4 Constructors for decision trees} *)
+
+(** A constructor is the representation of a symbol along with the number of
+    arguments to which it is applied. *)
+type constructor = { c_mod : string list
+                   ; c_sym : string
+                   ; c_ari : int }
+
+(** Functional map on constructors.  Special processing for small maps. *)
+module type constructor_mapping =
+sig
+  (** Type of keys. *)
+  type key = constructor
+
+  (** Type of a mapping. *)
+  type 'a t
+
+  (** The empty map. *)
+  val empty : 'a t
+
+  (** [is_empty m] returns whether [m] is empty; more efficient than comparing
+      to [empty] (avoids a conversion). *)
+  val is_empty : 'a t -> bool
+
+  (** [add k e m] adds element [e] with key [k] to mapping [m]. *)
+  val add : key -> 'a -> 'a t -> 'a t
+
+  (** [bindings m] returns an associative list [(k, e)] with [k] key and [e]
+      element. *)
+  val bindings : 'a t -> (key * 'a) list
+
+  (** [find_opt k m] returns [Some(e)] with [e] the element mapped by [k]
+      through [m] and [None] if [k] is not in [m]. *)
+  val find_opt : key -> 'a t -> 'a option
+
+  (** [iter f m] iters function [f] on mapping [m]. *)
+  val iter : (key -> 'a -> unit) -> 'a t -> unit
+
+  (** [map f m] maps [f] on mapping m. *)
+  val map : ('a -> 'b) -> 'a t -> 'b t
+end
+
+module ConsMap : constructor_mapping =
+struct
+
+  (** [cons_compare c d] is a comparison function for constructors; more
+      efficient than the pervasive. *)
+  let cons_compare : constructor -> constructor -> int = fun ca cb ->
+    let scomp = String.compare ca.c_sym cb.c_sym in
```

Comment:
You can instead do:
```
match String.compare ... with
| 0 ->
  begin
    match Int.compare ... with
    ...
  end
| x -> x
```

Reply:
It is probably worth hashconsing strings to replace String.compare by Int.compare.

Reply:
I guess it is already done by `String.compare`.  I considered that `String.compare` is as optimised as it could to compare `strings`.

Reply:
I don't think so but this is an optim that can be done later.

Reply:
Indeed you're right, `String.compare` calls `Stdlib.compare`